Monday, 25 March 2013 
====================================

Trying to figure out how to handle errors... All messages, actually....

In my controller, I try to go to a page that doesn't exist. This should bring up a 404 error with some message text (eg, "Page doesNotExist was not found"). In some cases, I'll have custom errors. (eg, "Bad JSON string", etc.). I can just return empty objects, but I'd like to pass back something that indicates what went wrong.... Essentially, I'd like this:

    Return a 404 code in the header
    {"errors":["item doesNotExist was not found"]}
    {"errors":["vendor doesNotExist was not found for item 2"]}
    {"errors":["incorrectly formatted JSON string submitted","Unable to parse text"]}

On success, it's currently just passing back the data. This is probably OK for now, but in the future, I may want to have more control. I could see a case where I'm doing paging, and wanting to do something like this:

    {"data":[...several returned objects here...],
     "offset": 100,
     "limit": "20"}

Maybe there are better ways, though, and I'll come across them in due course.



Error Message Handling
-----------------------
From http://stackoverflow.com/questions/14455112/distinguish-between-validation-errors-in-laravel

Get an instance of a MessageBag object, that you can run through with a foreach loop:

    $messages = $validator->messages(); // Where $validator is your validator instance.
    $messages = $messages->all()

    foreach ($messages as $message) {
        print $message;
    }

You can also get error messages for a given attribute:

    $errors = $validation->errors->get('email');

and then loop through the errors

    foreach ($errors as $error) { print $error; }

or get all the error messages

    $errors = $validation->errors->all();

then loop through the error messages

    foreach ($errors as $error) { print $error; }




Tuesday, 25 March 2013 
======================================================

Trying to figure out how to mock the database layer. 

There's some code here:
http://net.tutsplus.com/tutorials/php/testing-like-a-boss-in-laravel-models/

I don't want to use a table, though, I just want it to be mocked, and to return the mock values.

Hmmmm... This seems to be beyond my capabilities right now. Let's keep moving in some other way.

Let's just use an in-memory table.

Created file app/config/database.php:

    <?php
     
    return array(
     
        'default' => 'sqlite',
        'connections' => array(
            'sqlite' => array(
                'driver'   => 'sqlite',
                'database' => ':memory:',
                'prefix'   => ''
            ),
        )
    );

In TestCase.php, I'm including this code:

    public function loadDatabase()
    {
        Artisan::call('migrate');
        $this->seed();        
    }

And, in ItemTest.php, I'm including this:

    public function testFindItem()
    {
        $this->loadDatabase();
        $item = new Item;

        // // var_dump($item);
        var_dump($item->find(1));
    }

It seems to be getting stuck in an infinite loop, though. :-(

Even when I pull everything directly into the ItemTest function...

Hmmm.... Might as well do something else for a little bit...

composer update
Now everything is broken...
git clone -b develop https://github.com/laravel/laravel.git
(pull the latest skeleton)

Copied files to my working branch...
And now the infinite loop is magically gone, and we're working again. Yay!

Can we search for parameters?
Where would we do that?

I'd like to change the find function to search (or could include a 'search' function). Basically, I want the model to be a black box where search parameters can be dropped, and results are returned.

When we do a search, it's case sensitive.

This will not get anything. It's putting in an "and" by default

    return parent::where('name','=','Pencils')
        ->where('sku', '=', '2383243XDF');

This will return both pencils and windex:

    return parent::where('name','=','Pencils')
        ->orWhere('sku', '=', '2383243XDF');

This will just return windex (because it's case sensitive):

    return parent::where('name','=','pencils')
        ->orWhere('sku', '=', '2383243XDF');

It is possible to do a case insensitive match, but a bit ugly...
http://forums.laravel.io/viewtopic.php?id=1935

    public function get_index()
    {
        $firstname = Str::lower(Input::get('first_name',''));
        $lastname = Str::lower(Input::get('last_name',''));
        $c = $u->contacts();  // Referencing the relationship...
        // $c->select('*'); // Drop this
        if($firstname){$c->where(DB::raw('lower(`first_name`)', '=', $firstname;}
        if($lastname){$c->where(DB::raw('lower(`last_name`)', '=', $lastname;}
    }



Wednesday, 26 March 2013 
======================================================

Updated laravel with the latest development commits.

    git clone -b develop https://github.com/laravel/laravel.git laravel4
    cp -r laravel4/* ../kdb_shopping/api/
    composer update

git status shows several files have been modified:

    modified:   app/config/database.php
    modified:   app/config/mail.php
    modified:   app/controllers/BaseController.php
    modified:   app/database/seeds/DatabaseSeeder.php
    modified:   app/models/User.php
    modified:   app/routes.php
    modified:   app/tests/TestCase.php
    modified:   composer.json
    modified:   phpunit.xml

Since my project is being tracked, and I was up-to-date before copying the new laravel skeleton, I can just do this:

    git reset --hard HEAD
    composer dump-autoload

phpunit is complaining, though, with a permission denied error. 

    sudo chown www-data app/storage/{cache,logs,meta,sessions,views}

Good to go.


Sublime Keymap
---------------

Changed my sublime keymap to this:

[
    { "keys": ["ctrl+shift+;"], "command": "run_macro_file", "args": {
            "file": "Packages/User/EOL_Semicolon.sublime-macro"
        } 
    },
    { "keys": ["ctrl+alt+n"], "command": "advanced_new_file"},
    { "keys": ["ctrl+shift+a"], "command": "select_all" },
    { "keys": ["ctrl+a"], "command": "move_to", "args": {"to": "bol", "extend": false} },
    { "keys": ["ctrl+e"], "command": "move_to", "args": {"to": "eol", "extend": false} }
]


Item
-------------

More testing for the Item object. If I use the 'like' operator, rather than '=', it's case insensitive. Yay! I was after that, anyway.

Here's my search function, which works nicely:

    public static function search($findWhat)
    {
        // If we were passed an integer, find the primary key
        if(is_numeric($findWhat)) {
            return parent::find($findWhat);            
        } 

        // search for submitted parameters
        $params = array();
        parse_str($findWhat, $params);
        $foundItems = parent::from('items');
        foreach($params as $key => $value) {
            $foundItems->where($key, 'like', '%'.$value.'%', 'and');
        }
        if (count($foundItems->get())==1) {
            return $foundItems->first();
        }
        return $foundItems->get();
    }

Next issue: If (manually, in the browser) I send a request to items?name=pencil, I get every item. If I use items/name=pencil, I just get the pencils. How do I include a query string? 

In the index function, I can look for input...

    public function index()
    {
        if (count(Input::all())==0)
            return $this->items->all();

        return $this->show(Input::all());
    }

Also, how do I automatically test for it?
If I use $this->get('items/name=pencil'), it throws an error:

    Symfony\Component\HttpKernel\Exception\NotFoundHttpException



Thursday, 27 March 2013 
======================================================

How can I call some urls with query parameters in tests?

I have some urls: 

    http://api.shop/items?name=x    (routes through index)
    http://api.shop/items/name=x    (routes through show via resource routing)
    http://api.shop/items/?name=x   (not found exception)

In my test function, this works great:

    return $this->call('GET', 'items');         // returns all items
    return $this->call('GET', 'items/1');       // returns item 1

But these fail in various ways:

    return $this->call('GET', 'items/?name=x');               not found exception
    return $this->call('GET', 'items/name=x');                not found exception
    return $this->call('GET', 'items?name=x');                all items
    return $this->call('GET', 'items', array('name'=>'x'));   all items
    return $this->call('GET', 'items', array('name=x'));      all items
    return $this->call('GET', 'items', array('name'=>'x'));   all items

There are some samples in the code:

vendor/symfony/http-foundation/Symfony/Component/HttpFoundation/Tests/RequestTest.php:
$request = Request::create('http://test.com/foo?bar=foo', 'GET', array('bar' => 'baz'));

vendor/symfony/http-foundation/Symfony/Component/HttpFoundation/Tests/RequestTest.php:
$request = Request::create('http://thomas:pokemon@example.net:8080/?foo=bar', 
    'GET', array(), array(), array(), array());

vendor/symfony/http-kernel/Symfony/Component/HttpKernel/Tests/ClientTest.php:
$client->request('GET', 'http://www.example.com/?parameter=http://google.com');

vendor/laravel/framework/tests/Pagination/PaginationEnvironmentTest.php:
$request = Illuminate\Http\Request::create('http://foo.com?page=2', 'GET');

vendor/laravel/framework/tests/Http/HttpRequestTest.php:
$request = Request::create('http://foo.com/foo/bar?name=taylor', 'GET');


It looks like they're failing because I'm putting several assertions in the same test. When I separate them out, they work fine.

This works:

    public function testSearchForItemByName()
    {
        $this->prepareForTests();
        
        $json = $this->getJSON('items/name=x');
        $this->assertRecordFound($json, 'name', 'windex', 
            'should find Windex for items/name=w'); 
        $this->assertRecordNotFound($json, 'name', 'pencil', 
            'should not find pencil for items/name=w'); 
    }

    public function testSearchForItemByNameInQuery()
    {
        $this->prepareForTests();

        $json = $this->getJSON('items?name=x');
        $this->assertRecordFound($json, 'name', 'windex', 
            'should find Windex for items?name=w'); 
        $this->assertRecordNotFound($json, 'name', 'pencil', 
            'should not find pencil for items?name=w'); 
    }

This doesn't:

    public function testSearchForItem()
    {
        $this->prepareForTests();
        
        $json = $this->getJSON('items/name=x');
        $this->assertRecordFound($json, 'name', 'windex', 
            'should find Windex for items/name=w'); 
        $this->assertRecordNotFound($json, 'name', 'pencil', 
            'should not find pencil for items/name=w'); 

        $json = $this->getJSON('items?name=x');
        $this->assertRecordFound($json, 'name', 'windex', 
            'should find Windex for items?name=w'); 
        $this->assertRecordNotFound($json, 'name', 'pencil', 
            'should not find pencil for items?name=w'); 
    }


Laravel Environment
--------------------

Determine what environment we're running in, with:

    app()->make('env')
    App::make('env')

The environment defaults to 'production'
Other options are detected in bootstrap/start.php:

    $env = $app->detectEnvironment(array(
        'dev' => array('machine_id','*.shop'),  // machine name and/or host
    ));

It's also changed in TestCase (base class for all tests):

    public function createApplication()
    {
        $unitTesting = true;
        $testEnvironment = 'testing';
        return require __DIR__.'/../../bootstrap/start.php';
    }

I can set up a separate test environment, called foo, like so:

    public function createApplication()
    {
        $unitTesting = true;
        $testEnvironment = 'testing';
        if(in_array('foo', $_SERVER['argv'])) {
            $testEnvironment = 'foo';
        }
        return require __DIR__.'/../../bootstrap/start.php';
    }

It will read all configuration settings from the foo directory.
Run phpunit with the foo configuration by entering:

    $ phpunit -d foo

We can also set other command line parameters, so this works:

    $ phpunit -d foo -d bar -d etc

I created a production folder. Now, dev is the default environment, and there are different settings for production, testing, and foo (staging, etc.)


Debugging
---------------

Installed SublimeXDebug
    Shift+f8:        Open XDebug quick panel
    f8:              Open XDebug control quick panel when debugger is connected
    Ctrl+f8:         Toggle breakpoint
    Ctrl+Shift+f5:   Run to next breakpoint
    Ctrl+Shift+f6:   Step over
    Ctrl+Shift+f7:   Step into
    Ctrl+Shift+f8:   Step out

It doesn't stop on a breakpoint in either a browser or a phpunit test, so far.

Adding xdebug helper to chrome
Interesting... with the xdebug helper, when I start or stop debugging in sublime, it does something in the browser. 


Laravel Validation
--------------------

The validator object needs to be sent an array, or it will throw an error (eg, if a ParameterBag is sent). I can convert those to an array in the controller, but I think I want the model layer to handle all data-related issues...




Sunday, 31 March 2013 
======================================================

Unit Testing for Exceptions
----------------------------

Trying to unit test my BaseController class. Basically, it just returns exceptions and appropriate messages. 

We can check to see if we're getting an exception with a doc block:

    /**
    * @expectedException \Symfony\Component\HttpKernel\Exception\NotFoundHttpException
    */
    public function testNotFoundWorks()
    {
        $bc = new BaseController();
        $response = $bc->notFound('test');
    }
    
This works to let us know that a NotFoundHttpException was sent, but it doesn't give us more information.

    public function testNotFoundWorks()
    {
        $bc = new BaseController();
        try {
            $response = $bc->notFound('test');
        } catch (NotFoundHttpException $e) {
            var_dump($e->getStatusCode());
            var_dump($e->getHeaders());
        }
    }
    
This shows me the status code and the headers, but still not the message.

    var_dump($e->getMessage()) returns an empty string...
    
If I were to call it differently, it would return some text. This is a revised way to call it:

    try{
        throw new NotFoundHttpException('test');
    } catch (NotFoundHttpException $e) {
        var_dump($e->getMessage());
    }

I suspect there's something in Laravel, though.
How to get at it...

What if we were to put in a unique phrase:

    try {
        $response = $bc->notFound('stuffForJoel');
    } catch (NotFoundHttpException $e) {
        var_dump($e);
    }

Redirect the output to a text file, and grep for that string... It's there. Let's see the context in which it's found...

Hmmmm.... Only in the trace call stack... not so helpful.

Let's go the other way. I'm returning a Response object from the function. 

  Illuminate\Http\Response
  \Symfony\Component\HttpFoundation\Response

Ahhh... it's just going through the controller, the magic method __call is throwing the exception.

Right. It's calling static methods. Use :: instead of ->
... and they're protected, so they can only be called by descendents. We'll need to instanciate a descendent class.

